File: lib/nats_transport.go
package frugal

import (
	"git.apache.org/thrift.git/lib/go/thrift"
	"github.com/nats-io/nats"
)

type natsTransportFactory struct {
	conn *nats.Conn
}

func NewNATSTransportFactory(conn *nats.Conn) TransportFactory {
	return &natsTransportFactory{conn: conn}
}

func (n *natsTransportFactory) GetTransport(topic string) Transport {
	return NewNATSTransport(n.conn, topic)
}

type natsTransport struct {
	thriftTransport thrift.TTransport
}

func NewNATSTransport(conn *nats.Conn, subject string) Transport {
	return &natsTransport{newNATSThriftTransport(conn, subject)}
}

func (n *natsTransport) Subscribe() error {
	return n.thriftTransport.Open()
}

func (n *natsTransport) Unsubscribe() error {
	return n.thriftTransport.Close()
}

func (n *natsTransport) ThriftTransport() thrift.TTransport {
	return n.thriftTransport
}

func (n *natsTransport) ApplyProxy(transportFactory thrift.TTransportFactory) {
	n.thriftTransport = transportFactory.GetTransport(n.thriftTransport)
}

File: lib/src/nats_transport.dart
part of frugal;

class NatsTransportFactory implements FTransportFactory {
  Nats client;

  NatsTransportFactory(this.client);

  FTransport getTransport() => new FNatsTransport(this.client);
}


class FNatsTransport implements FTransport {
  TTransport tTransport;
  NatsThriftTransport nTransport;

  Stream get signalRead => nTransport.signalRead;
  Stream get error => nTransport.error;

  FNatsTransport(Nats client) {
    var tr = new NatsThriftTransport(client);
    tTransport = tr;
    nTransport = tr;
  }

  Future subscribe(String subject) {
    nTransport.setSubject(subject);
    return nTransport.open();
  }

  Future unsubscribe() {
    return nTransport.close();
  }

  void preparePublish(String subject) {
    nTransport.setSubject(subject);
  }

  TTransport thriftTransport() => tTransport;

  Future applyProxy(TTransportFactory transportFactory) async {
    tTransport = await transportFactory.getTransport(tTransport);
  }
}
